Python/Product/Analysis/Interpreter/Ast/AstAnalysisWalker.cs

```diff
@@ -216,39 +230,94 @@ private IMember Clone(IMember member)
-                            _log?.Log(TraceLevel.Warning, "UndefinedImport", modName, name);
+                        var mem = mod.GetMember(Scope.Context, member);
+                        if (mem == null) {
+                            if (WarnAboutUndefinedValues) {
+                                _log?.Log(TraceLevel.Warning, "UndefinedImport", modName, member);
+                            }
+                            mem = new AstPythonConstant(_unknownType, ((mod as ILocatedMember)?.Locations).MaybeEnumerate().ToArray());
+                        } else if (mem.MemberType == PythonMemberType.Unknown && warnAboutUnknownValues) {
+                            _log?.Log(TraceLevel.Warning, "UnknownImport", modName, member);
                         }
-                        Scope.SetInScope(member, mem);
+                        Scope.SetInScope(member, mem, scope: scope);
                         (mem as IPythonModule)?.Imported(Scope.Context);
-                        if (isTyping) {
-                            _typingMembers.Add(member);
-                        }
                     }
                 } else {
                     IMember mem;
-                    if (mod.IsLoaded) {
-                        mem = mod.GetMember(Scope.Context, name.Key) ??
-                            new AstPythonConstant(_unknownType, GetLoc(name.Value));
-                        if (mem.MemberType == PythonMemberType.Unknown && WarnAboutUndefinedValues) {
-                            _log?.Log(TraceLevel.Warning, "UndefinedImport", modName, name);
+                    if (mod is AstNestedPythonModule m && !m.IsLoaded) {
+                        mem = new AstNestedPythonModuleMember(name.Key, m, Scope.Context, GetLoc(name.Value));
+                    } else {
+                        mem = mod.GetMember(Scope.Context, name.Key);
+                        if (mem == null) {
+                            if (WarnAboutUndefinedValues) {
+                                _log?.Log(TraceLevel.Warning, "UndefinedImport", modName, name);
+                            }
+                            mem = new AstPythonConstant(_unknownType, GetLoc(name.Value));
+                        } else if (mem.MemberType == PythonMemberType.Unknown && warnAboutUnknownValues) {
+                            _log?.Log(TraceLevel.Warning, "UnknownImport", modName, name);
                         }
                         (mem as IPythonModule)?.Imported(Scope.Context);
-                    } else {
-                        mem = new AstNestedPythonModuleMember(name.Key, mod, Scope.Context, GetLoc(name.Value));
-                    }
-                    Scope.SetInScope(name.Value.Name, mem);
-                    if (isTyping) {
-                        _typingMembers.Add(name.Value.Name);
                     }
+                    Scope.SetInScope(name.Value.Name, mem, scope: scope);
```

Comment:
OK for `mem` to be null?

Reply:
It shouldn't ever be null, but yes it's okay (it means to remove the value from the current scope).